File: java-algorithm/src/main/java/com/leetcode/Problem1663.java
package com.leetcode;

/**
 * 1663. Smallest String With A Given Numeric Value
 * https://leetcode.com/problems/smallest-string-with-a-given-numeric-value/
 */
public class Problem1663 {
    public String getSmallestString(int n, int k) {
        int countZ = (k - n) / 25;
        int countA = (k - n) % 25 == 0 ? n - countZ : n - countZ - 1;
        StringBuilder sb = new StringBuilder();
        sb.append("a".repeat(countA));
        if ((k - n) % 25 > 0) {
            sb.append((char)('a' + (k - n) % 25));
        }
        sb.append("z".repeat(countZ));
        return sb.toString();
    }
}

File: python-algo/leetcode/problem_1663.py
"""1663. Smallest String With A Given Numeric Value
https://leetcode.com/problems/smallest-string-with-a-given-numeric-value/
"""


class Solution:
    def getSmallestString(self, n: int, k: int) -> str:
        z, rem = divmod(k, 26)
        ans = ['z'] * z
        if rem > 0:
            ans.insert(0, chr(96 + rem))
        if len(ans) == n:
            return ''.join(ans)
        diff = n - len(ans)
        prefix = ''
        i = 0
        while diff > 0:
            x = min(ord(ans[i]) - 97, diff)
            ans[i] = chr(ord(ans[i]) - x)
            prefix += 'a' * x
            diff -= x
            i += 1
        return prefix + ''.join(ans)
